pkg/auth/globalrole.go

```diff
@@ -0,0 +1,85 @@
+package auth
+
+import (
+	"fmt"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	controllerv3 "github.com/rancher/webhook/pkg/generated/controllers/management.cattle.io/v3"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+type GlobalRoleResolver struct {
+	roleTemplateResolver *RoleTemplateResolver
+	globalRoles          controllerv3.GlobalRoleCache
+}
+
+const ownerRT = "cluster-owner"
+
+var adminRoles = []string{"admin", "restricted-admin"}
+
+// NewRoleTemplateResolver creates a newly allocated RoleTemplateResolver from the provided caches
+func NewGlobalRoleResolver(roleTemplateResolver *RoleTemplateResolver, grCache controllerv3.GlobalRoleCache) *GlobalRoleResolver {
+	return &GlobalRoleResolver{
+		roleTemplateResolver: roleTemplateResolver,
+		globalRoles:          grCache,
+	}
+}
+
+// GlobalRulesFromRole finds all rules which apply globally - meaning valid for escalation checks at the cluster scope
+// in the local cluster.
+func (g *GlobalRoleResolver) GlobalRulesFromRole(gr *v3.GlobalRole) []rbacv1.PolicyRule {
+	// no rules on a nil role
+	if gr == nil {
+		return nil
+	}
+	return gr.Rules
+}
+
+// ClusterRulesFromRole finds all rules which this gr gives on downstream clusters
+func (g *GlobalRoleResolver) ClusterRulesFromRole(gr *v3.GlobalRole) ([]rbacv1.PolicyRule, error) {
+	if gr == nil {
+		return nil, nil
+	}
+	// admin and restricted admin are treated like they are owners of all downstream clusters
+	// but they don't get the same field because this would duplicate legacy logic
+	for _, name := range adminRoles {
+		if gr.Name == name {
+			templateRules, err := g.roleTemplateResolver.RulesFromTemplateName(ownerRT)
+			if err != nil {
+				return nil, fmt.Errorf("unable to resolve cluster-owner rules: %w", err)
+			}
+			return templateRules, nil
+		}
+	}
+	var rules []rbacv1.PolicyRule
+	for _, inheritedRoleTemplate := range gr.InheritedClusterRoles {
+		templateRules, err := g.roleTemplateResolver.RulesFromTemplateName(inheritedRoleTemplate)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get cluster rules for roleTemplate %s: %w", inheritedRoleTemplate, err)
+		}
+		rules = append(rules, templateRules...)
+	}
+	return rules, nil
+}
+
+// GetGlobalRole allows the caller to retrieve a global role by name
+func (g *GlobalRoleResolver) GetGlobalRole(name string) (*v3.GlobalRole, error) {
+	return g.globalRoles.Get(name)
+}
```

Comment:
Personally speaking, I don't like how that structure exposes internal resources for usage by other packages. I'd prefer that it offered an interface for performing specific operations that someone would want the cache for. That being said, I'll adjust it for the sake of consistency since this isn't that big of a deal.